### PR TITLE
Add wildcard configuration option for URI routes

### PR DESCRIPTION
**Pull request by GlaciErrDev**
The main idea behind URI route wildcard support is to avoid creating
  a ton of duplicate configurations for many services with long routes.

### CODE REVIEW

**Comment by auhlig** [comment]
> The main idea behind URI route wildcard support is to avoid creating
> a ton of duplicate configurations for many services with long routes.

Just to outline the reasoning for not introducing this initially:
I see the point of introducing wildcards for this to make configuration less verbose. 
However, wildcards bear the risk of implicitly and unintentionally introducing rate limits on routes.
An explicit configuration makes it easier to spot which rate limits are set for a specific route.
At least initially the set of routes that should be rate-limited was quite manageable - mostly create/update involving a lot of moving part in the resp. backends (server create, etc.).

**Comment by galkindmitrii** [comment]
> > The main idea behind URI route wildcard support is to avoid creating
> > a ton of duplicate configurations for many services with long routes.
> 
> Just to outline the reasoning for not introducing this initially:
> I see the point of introducing wildcards for this to make configuration less verbose.
> However, wildcards bear the risk of implicitly and unintentionally introducing rate limits on routes.
> An explicit configuration makes it easier to spot which rate limits are set for a specific route.
> At least initially the set of routes that should be rate-limited was quite manageable - mostly create/update involving a lot of moving part in the resp. backends (server create, etc.).

Hi Arno,
Thanks for the input.
As I see it, there is also the opposite risk of certain endpoints not covered by the limits by mistake.
Nova has something like 50+ routes and this number might change as release upgrade comes in.
For the specific routes that require different limits it should still be possible to set those, but I'd expect the wildcards to be used most of the time.

**Comment by auhlig** [comment]
> Nova has something like 50+ routes and this number might change as release upgrade comes in.
But do all of them need rate limiting? The important set was a handful of "expensive" routes IMHO.

Feel free to add wildcard support nevertheless. It's definitely useful. Just wanted to share my thoughts on.
If it's not already doing that: Maybe add a log line to make it clear which wildcard target type uri and configured rate limit was used.

File: rate_limit/provider.py [hunk 1]
```diff
@@ -97,12 +103,51 @@ def get_local_rate_limits(self, scope, action, target_type_uri, **kwargs):
...
+        Get the target type URI rate limits from wildcard pattern.
+
+        :param target_type_uri: the target type URI of the request
+        :return: target type uri ratelimits if exists
+        """
+
+        ttu_ratelimits = []
+        pattern_list = [
+            lr_key for lr_key in ratelimits
+            if lr_key.endswith('*')
```
Comment (chuan137) [inline rate_limit/provider.py:128]: Does this means that rule ending with "\*" is only allowed? Is "/foo/*/bar/*" allowed?

File: rate_limit/provider.py [hunk 2]
```diff
@@ -97,12 +103,51 @@ def get_local_rate_limits(self, scope, action, target_type_uri, **kwargs):
...
+        Check if a URI matches to one of the patterns
+
+        :param uri: URI to check if it matches to one of the patterns
+        :param pattern_list : patterns to match against the URI
+        :return: True if path matches a pattern of the list and
+                 pattern as key for self.local_ratelimits.
+        """
+
+        for pattern in pattern_list:
+            if fnmatch.fnmatch(uri, pattern):
```
Comment (chuan137) [inline rate_limit/provider.py:147]: If only pattern ending with "\*" is expected, why not check the ttu with string's startswith function? It avoids additional dependent lib and might be a performance advantage.

**Review by chuan137 — COMMENTED** [review]
Please see my inline comments

File: rate_limit/provider.py (hunk 1 above)
Comment (GlaciErrDev) [reply rate_limit/provider.py:128]: Yes, only allowed for rules ending with "*". Discussed this with @galkindmitrii

Comment (chuan137) [reply rate_limit/provider.py:128]: okay, in that case, I think it makes sense to validate the pattern somewhere once

**Pull request merged**